### PR TITLE
ussologin: change TokenStore interface

### PR DESCRIPTION
**Pull request by mhilton**
Add a name parameter to the TokenStore interface. Also create a directory
based TokenStore.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/idmclient/40/
Test PASSed.

File: ussologin/ussologin.go [hunk 1]
```diff
@@ -157,5 +162,60 @@ func (f *FileTokenStore) Get() (*usso.SSOData, error) {
...
+// for storage. All token files will be have the name prefix + tokenName.
+func NewDirTokenStore(path, prefix string) *DirTokenStore {
+	return &DirTokenStore{
+		path:   path,
+		prefix: prefix,
+	}
+}
+
+// Put implements TokenStore.Put by writing the token to the
+// DirTokenStore's directory.
```
Comment (rogpeppe) [inline ussologin/ussologin.go]: // It creates the directory and any parent directories if necessary.

?

File: ussologin/ussologin.go [hunk 2]
```diff
@@ -157,5 +162,60 @@ func (f *FileTokenStore) Get() (*usso.SSOData, error) {
...
+	return nil
+}
+
+// Get implements TokenStore.Get by
+// reading the token from the DirTokenStore's directory.
+func (s *DirTokenStore) Get(name string) (*usso.SSOData, error) {
+	data, err := ioutil.ReadFile(filepath.Join(s.path, s.prefix+name))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
```
Comment (rogpeppe) [inline ussologin/ussologin.go]: I'm not sure it's right that this returns nil if the token's not found.
I think it should return some kind of not-found error
(and that could be documented in the tokenstore interface).

File: ussologin/ussologin.go [hunk 3]
```diff
@@ -115,10 +117,13 @@ func LoginWithToken(client *http.Client, ussoAuthUrl string, tok *usso.SSOData)
...
-	// Put stores an Ubuntu SSO OAuth token.
+	// Put stores an Ubuntu SSO OAuth token. The token should be
+	// retireved from the store using Get(tok.TokenName).
 	Put(tok *usso.SSOData) error
-	// Get returns an Ubuntu SSO OAuth token from store
-	Get() (*usso.SSOData, error)
+
+	// Get returns an Ubuntu SSO OAuth token, identified by name,
+	// from store
+	Get(name string) (*usso.SSOData, error)
```
Comment (rogpeppe) [inline ussologin/ussologin.go]: I'm not sure I'm keen on this. I think it works better as a straight store for a token with no regard to what that token is. The caller can check if the retrieved token matches their expectation if it likes.

File: ussologin/ussologin.go [hunk 4]
```diff
@@ -157,5 +162,60 @@ func (f *FileTokenStore) Get() (*usso.SSOData, error) {
 	if err := json.Unmarshal(data, &tok); err != nil {
 		return nil, errgo.Notef(err, "cannot unmarshal token")
 	}
+	if tok.TokenName != name {
+		return nil, nil
```
Comment (rogpeppe) [inline ussologin/ussologin.go]: Again, returning (nil, nil) seems wrong to me. Unless specifically documented in the interface, the returned token should be non-nil iff err == nil.

**Comment by rogpeppe** [comment]
I'm not that keen on this as it is, I'm afraid. It seems to confuse the tokenstore interface
contract without much good reason that I can see.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/idmclient/41/
Test PASSed.

File: ussologin/ussologin.go [hunk 5]
```diff
@@ -140,6 +142,10 @@ func (f *FileTokenStore) Put(tok *usso.SSOData) error {
 	if err != nil {
```
Comment (rogpeppe) [inline ussologin/ussologin.go]: Please mention in the docs that the directory is created if it
doesn't exist.

**Comment by rogpeppe** [comment]
LGTM modulo one doc suggestion.

**Comment by mattyw** [comment]
LGTM after the doc change

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/idmclient/42/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/idmclient/43/
Test PASSed.

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/idmclient-merge

**Pull request merged**